drivers/serial/uart_ns16550.c

```diff
@@ -505,11 +521,15 @@ static int uart_ns16550_poll_in(struct device *dev, unsigned char *c)
 static void uart_ns16550_poll_out(struct device *dev,
 					   unsigned char c)
 {
+	k_spinlock_key_t key = k_spin_lock(&DEV_DATA(dev)->lock);
+
 	/* wait for transmitter to ready to accept a character */
 	while ((INBYTE(LSR(dev)) & LSR_THRE) == 0) {
```

Comment:
Relevant to the discussion earlier: this is a potential deadlock right here, though unlikely to be the source of the hang mentioned above.  This is going to spin with the lock held waiting for the hardware FIFO to drain.  But some uarts (at least in theory) have hardware implementations of flow control, which means that the device on the other side is capable of stalling the FIFO at will and causing a hang locally.

This while need to go outside a single block that does (atomically) "can I send a byte?  if so, send it and break out of the loop".   And it's probably good practice to put a busy wait of ~1 byte transfer time (87us at 115200bps) to avoid spamming the lock and/or banging the hardware too fast.

Reply:
ah, wonderful. makes sense

Reply:
That sounds similar to what's going on in https://github.com/zephyrproject-rtos/zephyr/pull/25064/files?  I tried that approach but did not resolve the hang. Still investigating

Reply:
@andyross Interesting. There are a few functions that seem to fit this scenario...perhaps I need to make the same adjustment with a while loop outside a single block that does the checking/doing, yeah?  `uart_ns16550_poll_out`, `uart_ns16550_poll_in`, `uart_ns16550_fifo_fill`, and `uart_ns16550_fifo_read`

Reply:
@jenmwms: Mind that there's a huge difference between poll_out/poll_in vs fifo_fill/fifo_read. Please refer to the docstrings: https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/uart.h#L703 . To quote import points:

> This function is expected to be called from UART
> interrupt handler (ISR)

> Result of calling this function not from an ISR is undefined

> Likewise, *not* calling this function
> from an ISR if uart_irq_tx_ready() returns true may lead to
> undefined behavior

As pointed by @andyross, normally, one doesn't use locking in ISRs. Thus, a solution suitable for poll_out/poll_in, wouldn't be suitable for fifo_fill/fifo_read, and vice versa, solution for fifo_fill/fifo_read might be not the best for poll_out/poll_in.

Reply:
@pfalcon Thanks for the info and link, it is helpful. I'll keep that in mind as I keep working on this.

Reply:
@jenmwms Almost always yes.  The point to a spinlock[1] is that it needs to enclose a critical section with firmly bounded runtime.  It's not like you can never use a "while" in there syntactically, but never in a circumstance where you aren't 100% sure exactly how long it will be until you unlock.

[1] In production code.  In the test suite we do all kinds of abusive things to locking paradigms to isolate stuff we want to test.

@pfalcon: You still need locking in ISRs, though the rules get more complicated.   Almost all Zephyr platforms support nested interrupts, so it's rarely actually true that you know you won't be interrupted.  And the driver in question is used on SMP systems, where you can have an interrupt racing against code on another CPU.

Reply:
@andyross, Right, someone *may* need locking in ISRs, and the rules for deal with those cases are complicated, so I didn't try to mention them in my comment.